svm/src/account_loader.rs

```diff
@@ -213,29 +213,30 @@ fn load_transaction_accounts<CB: TransactionProcessingCallback>(
             let account = if solana_sdk::sysvar::instructions::check_id(key) {
                 construct_instructions_account(message)
             } else {
-                let instruction_account = u8::try_from(i)
-                    .map(|i| instruction_accounts.contains(&&i))
-                    .unwrap_or(false);
                 let (account_size, mut account, rent) = if let Some(account_override) =
                     account_overrides.and_then(|overrides| overrides.get(key))
                 {
                     (account_override.data().len(), account_override.clone(), 0)
-                } else if let Some(program) = (!instruction_account && !message.is_writable(i))
-                    .then_some(())
-                    .and_then(|_| loaded_programs.find(key))
-                {
-                    callbacks
-                        .get_account_shared_data(key)
-                        .ok_or(TransactionError::AccountNotFound)?;
-                    // Optimization to skip loading of accounts which are only used as
-                    // programs in top-level instructions and not passed as instruction accounts.
-                    let program_account = account_shared_data_from_program(&program);
-                    (program.account_size, program_account, 0)
                 } else {
+                    let is_account_writable = message.is_writable(i);
                     callbacks
-                        .get_account_shared_data(key)
-                        .map(|mut account| {
-                            if message.is_writable(i) {
+                        .load_account_with(key, |account| {
+                            // only cache account that is not writable and not having program owners.
+                            !is_account_writable && !program_owners.contains(account.owner())
```

Comment:
Can you test how much performance changes if you remove the `!is_account_writable &&`? I suspect it would be barely measurable, as program management instructions are quite rare. Also, the writeability should not determine if it goes into the read cache or not.

Reply:
![image](https://github.com/anza-xyz/agave/assets/219428/1595d44d-42a1-4b3f-ae91-771baa65a5eb)
In the above graph, *red*, no writable check, shows better load time.
total replay time didn't change much.

Reply:
Can you run one more test with `.load_account_with(key, |account| false)`, effectively disabling the read caching altogether?

Reply:
sure. will do.

Reply:
![image](https://github.com/anza-xyz/agave/assets/219428/7411842b-53b6-468f-ab4a-3bc6211591c5)

Reply:
![image](https://github.com/anza-xyz/agave/assets/219428/3b2590b4-7923-4c17-95a0-5d93fe5c510c)

Reply:
![image](https://github.com/anza-xyz/agave/assets/219428/85c5db0e-da0f-4ac1-9879-e2b7cc178099)

Reply:
red: this PR. 
blue: disable read cache all together. 

14:25 is the time when the "blue" catches up to the network. 
Looks like load time is similar. blue (no read cache) seems to have less variance in load time.

Reply:
Ah. There is another difference.

"red" is running with high-low watermarks for read_cache:
 ` --accounts-db-read-cache-limit-mb 400,800`
"blue" node is running with default. 

But this shouldn't matter for blue because blue is not adding to read cache.

Reply:
A possible explanation is that loading from cache and doing lru is as expensive as loading from accounts-db?

Reply:
I am restarting "red" to use the default high-low water mark for read_cache. In this way, we will have the same read_cache config to compare.

Reply:
![image](https://github.com/anza-xyz/agave/assets/219428/3e26a9d8-26ce-4b89-906e-8df27b2f00c6)

![image](https://github.com/anza-xyz/agave/assets/219428/de7df226-8ad9-4e60-82af-7ac0c518e82a)

blue: disable read cache
red: enable read cache

blue is a little better on load time. 
blue is worse in cache miss and cache hit.

Reply:
So total replay just gets a little more variance with the read caching enabled.
But you did not disable it in both places where `load_account_with()` is used and missing program loading still does the read cached load right?

Reply:
Both places are disabled. 
The missing program loading already have the callback to return false.

Reply:
So where else is the read cache populated then? The blue read cache hits graph is not zero.

Reply:
program_loader.rs:106

Reply:
account_loader.rs:353

Reply:
There are many other places that load accounts and populate read cache besides transaction processor. 

1. programdata account
2. program owner account
3. sysvar account
4. vote account
5. feature account
6. buildin account
7. nounce account
8. fee account
9. withdraw/deposit

The following commit lists several places that we populate read cache.
https://github.com/anza-xyz/agave/pull/1398/commits/cc2bacc01e9d6166f928ddb2595a3fac8c335591

Some of them should be optimized to skip populating read cache. For example, vote account, and sysvar accounts and fee paying accounts. 

I think this is out of scope of this project. Probably, we need another project to reviews all the place that we load account and decide whether it is valuable to store the account into the cache.